Implemented PXB-2866 Make PXB 8.0.30 be able to backup 8.0.29 (and pr…

altmannmarcelo opened the pull request:
…ior) server

https://jira.percona.com/browse/PXB-2866

Adjusted redo log logic to populate start and end lsn of old redo log layout.
Changed redo_log.cc to use log_sys->m_files_ctx instead of creating a new Log_files_context each time. log_sys->m_files_ctx will already have the correct Log_files_ruleset (CURRENT or PRE_8_0_30) that can be used to identify if we are dealing with ib_logfile or #ib_redo layout.

Fixed --prepare warning about redo been created by unknown creator.

altmannmarcelo committed:
Implemented PXB-2866 Make PXB 8.0.30 be able to backup 8.0.29 (and prior) server

https://jira.percona.com/browse/PXB-2866

Adjusted redo log logic to populate start and end lsn of old redo log
layout.
Changed redo_log.cc to use log_sys->m_files_ctx instead of creating a new
Log_files_context each time. log_sys->m_files_ctx will already have the
correct Log_files_ruleset (CURRENT or PRE_8_0_30) that can be used to
identify if we are dealing with ib_logfile or #ib_redo layout.

Fixed --prepare warning about redo been created by unknown creator.

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1138,7 +1138,9 @@ static
     }
 
-    if (!file.contains(checkpoint_in_file.m_checkpoint_lsn)) {
+    if (IF_XB(log.m_files_ctx.m_files_ruleset >
+              Log_files_ruleset::PRE_8_0_30) &&
+        !file.contains(checkpoint_in_file.m_checkpoint_lsn)) {
```
why isn't applicable for pre 8030? checkpoint_lsn would be in the file detected above, isn't it?

altmannmarcelo replied:
on pre-8.0.30 we always have checkpoint marker on ib_logfile0, even if the LSN it points to is located on ib_logfile1. This causes the backup to fail in case we are currently writing on ib_logfile1.

satya-bodapati replied:
can you please add this as a comment? Maybe for versions < 8_0_30 we can check that the front file will always have checkpoint header?

satya-bodapati replied:
Another doubt I have is how will we find wrap around logic? because we always have two files loaded in loop. 
Unlike 8.0.30, the loaded files start_lsn will keep changing. In 8.0.30, the files start_lsn is fixed and new files appear.

altmannmarcelo replied:
@satya-bodapati logic explained as part of commit message (adjusted)
wrap around logic has been fixed at 8d8feb3

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -65,11 +65,8 @@ bool Redo_Log_Reader::find_start_checkpoint_lsn() {
-
   auto file_handle =
-      Log_file::open(log_files_ctx, checkpoint.m_checkpoint_file_id,
+      Log_file::open(log_sys->m_files_ctx, checkpoint.m_checkpoint_file_id,
```
where do we populate log_sy->m_files_ctx?

altmannmarcelo replied:
at the end'ish of log_sys_init in log0log.cc

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -132,12 +125,10 @@ static bool reopen_log_files() {
   log.m_files = std::move(files);
-  auto file = log.m_files.front();
-  log_encryption_read(log, file);
+  auto file = log.m_files.find(desired_lsn);
```
is the encryption information only in one of the ib_logfile* files?

side question: do we write encryption info in every ib_redo* file?

altmannmarcelo replied:
Adjusted to get it on the file contained the required LSN.

> side question: do we write encryption info in every ib_redo* file?

just the ones that have a checkpoint.

rahulmalik87 commented on storage/innobase/log/log0log.cc:
```diff
@@ -1546,7 +1546,8 @@ static dberr_t log_sys_handle_creator(log_t &log) {
     }
-  } else if (!str_starts_with(creator_name, "MySQL ")) {
+  } else if (!str_starts_with(creator_name, "MySQL ") &&
+             !str_starts_with(creator_name, "xtrabkup ")) {
```
we have added LOG_HEADER_CREATOR_PXB, so It is already handled.

rahulmalik87 commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -132,12 +125,10 @@ static bool reopen_log_files() {
-  log_encryption_read(log, file);
+  auto file = log.m_files.find(desired_lsn);
+  if (file == log.m_files.end()) return false;
+  log_encryption_read(log, *file);
```
Better to check db_err of log_encryption_read and return false, because it can fail if the file is removed by the server.  
Also we could only try to read encryption if redo_file is encrypted.

rahulmalik87 reviewed (changes requested):
How do we ensure wrap-around of redo files?

altmannmarcelo committed:
pre 8.0.30 backup

Adjusted lsn calculation on prepare when checkpoint fall into last
block.
PRE 8.0.30 were writting LOG_BLOCK_CHECKPOINT_NO on 8th byte of block
header, while on 8.0.30 we write LOG_BLOCK_EPOCH_NO. In cases where we
have two checkpoints faling in the same file (we dont re write the
header) when we reach to the next checkpoint and it falls into a new
block, LOG_BLOCK_CHECKPOINT_NO(which is stored as
recv_sys->scanned_epoch_no) will jump by two. In this case,
log_block_epoch_no_is_valid will identify the gap and will consider the
block as finished, without parsin it.

Adjusted consumer code to stop advancing after log_Status

altmannmarcelo committed:
Offset calculation:

[Xtrabackup] Log block checksum mismatch (block no 0 at lsn 18443264): expected 0, calculated checksum 3965168067 block epoch no: 0

Calculation of offsets have changed for 8.0.30+ as the files now carry
an ever increment ID which is used to determine what offset a specific
LSN belongs to (via file->offset(LSN) function). At pre 8.0.30 we had
to use log_files_real_offset_for_lsn.
Same function will have to be used by PXB, in fact we have it as part
of log0pre_8_0_30.cc. The only gotcha here is that it requires us
to at least know one offset for a given LSN. At previous redo log layout
we kept updating log.current_file_real_offset in order to have that
information. On new layout that information is not used anymore.
However, we still have (on old redo log format) LOG_CHECKPOINT_OFFSET,
which is exposed as m_checkpoint_offset from
log_pre_8_0_30::Checkpoint_header and can be used together with
m_checkpoint_lsn to compound the some file LSN & OFFSET.
As part of the LSN/OFFSET calibration, we require to adjust
find_start_checkpoint_lsn to switch between Log_checkpoint_location and
log_pre_8_0_30::Checkpoint_header objects to extract the required
information.

A new function with same name as recv_find_max_checkpoint was added to
log_pre_8_0_30 namespace to populate Checkpoint_header object.

altmannmarcelo committed:
Addressed review comments

Adjusted places that call log_encryption_read to read its return code
and abort in case of failure.

Removed handling for xtrabackup header creator.

altmannmarcelo committed:
Fixed pxb-2710-predict_memory.sh test case

Added a condition to only add new 8.0.30+ variable depending on the
version of backup server.
Also fixed a race condition when backup completes before the last
INSERT has the chance to commit. If the INSERT completes in between the
end of backup and before we run the record_db_state we might see
differences on verify_db_state as the last INSERT will be rolled back
by PXB.

altmannmarcelo committed:
Fixed all_files.sh test case

We no longer generate redo files to be carried over on copy-back.
Adjusted test case to not check for ib_logfile when running diff

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -162,8 +186,19 @@ lsn_t Redo_Log_Reader::read_log_seg(log_t &log, byte *buf, lsn_t start_lsn,
-
-    source_offset = file->offset(start_lsn);
+    if (log.m_files.ctx().m_files_ruleset == Log_files_ruleset::CURRENT) {
+      source_offset = file->offset(start_lsn);
```
@altmannmarcelo for pre 8030 files, cant we populate a file object with m_start_lsn ? Then same logic should work ?

satya-bodapati replied:
ie ib_logfile0 & ib_logfile1 map to Log_file objects with m_start_lsn. Is this possible or not?

altmannmarcelo replied:
@satya-bodapati as per our slack conversation and research, the problem comes when we remove redo logs and let the server re-create them. start_lsn is calculated different in this case on pre 8.0.30 and we cannot use `file->offset(lsn)` function to get the correct file offset.

satya-bodapati commented on storage/innobase/xtrabackup/test/t/pxb-2710-predict_memory.sh:
```diff
@@ -6,8 +6,13 @@ require_debug_pxb_version
 sync-binlog=0
-innodb-redo-log-capacity=512M
 "
+if is_server_version_higher_than 8.0.29
+then
+  MYSQLD_EXTRA_MY_CNF_OPTS="${MYSQLD_EXTRA_MY_CNF_OPTS:-""}
+  innodb-redo-log-capacity=512M
+  "
+fi
```
else 
innodb-log-file-size=512M?

altmannmarcelo replied:
it's not needed for pre 8.0.30 as we do not remove the files (wrap around) unless we actually write more than 48M*2

altmannmarcelo commented:
closed in favor of https://github.com/percona/percona-xtrabackup/pull/1377
